rclcpp/topics/minimal_subscriber/member_function_type_masquerading.cpp

```diff
@@ -0,0 +1,48 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs_conversions/std_string.hpp"
+
+using std::placeholders::_1;
+
+class MinimalSubscriber : public rclcpp::Node
+{
+public:
+  MinimalSubscriber()
+  : Node("minimal_subscriber")
+  {
+    subscription_ = this->create_subscription<std::string>(
+      "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
+  }
+
+private:
+  void topic_callback(const std::shared_ptr<std::string> msg) const
```

Comment:
I think as part of this work we may need to rethink what callback signatures we support.

For instance, I think either of these makes more sense than this signature (though it might be the right signature atm):

- `std::shared_ptr<const std::string> msg`
- `const std::shared_ptr<const std::string> & msg`
  - arguably just this one based on https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-const

In both cases we make clear it's the string, not the shared pointer that needs to be unchanged with in this function (because we might share it with other functions).

If the user wants to own a non-const copy of the message data (so they can mutate it) then we can let them make their own copy and we can give the const reference with this signature:

- `const std::string & msg`

If we're nice we could have some syntactic sugar to make the copy for them:

- `std::string msg` or
- `std::unique_ptr<std::string> msg`
  - they could upgrade this to a shared_ptr if they like

----

Nothing to do in this example, I just wanted to mention it while I was thinking about it. If we do change things we'd need to reflect them in this example, but also all the others. And we'd have to do a tick-tock on it (deprecation cycle).